src/feature/dirauth/dirvote.c

```diff
@@ -4198,31 +4199,62 @@ dirserv_get_bandwidth_for_router_kb(const routerinfo_t *ri)
   return bw_kb;
 }
 
-/** Helper for sorting: compares two routerinfos first by address, and then by
- * descending order of "usefulness".  (An authority is more useful than a
- * non-authority; a running router is more useful than a non-running router;
- * and a router with more bandwidth is more useful than one with less.)
+/** Helper for sorting: compares two ipv4 routerinfos first by ipv4 address,
+ * and then by descending order of "usefulness"
+ * (see compare_routerinfo_usefulness)
  **/
-static int
-compare_routerinfo_by_ip_and_bw_(const void **a, const void **b)
+STATIC int
+compare_routerinfo_by_ipv4(const void **a, const void **b)
 {
   routerinfo_t *first = *(routerinfo_t **)a, *second = *(routerinfo_t **)b;
+  // Don't use tor_addr_compare because it requires a tor_addr_t struct
+  // If addresses are equal, use other comparison criterions
+  if (first->addr == second->addr) {
+    return compare_routerinfo_usefulness(a,b);
+  } else {
+    // Otherwise, compare the addresses
+    if (first->addr < second->addr)
+      return -1;
+    else
+      return 1;
+  }
+}
+
+/** Helper for sorting: compares two ipv6 routerinfos first by ipv6 address,
+ * and then by descending order of "usefulness"
+ * (see compare_routerinfo_usefulness)
+ **/
+STATIC int
+compare_routerinfo_by_ipv6(const void **a, const void **b)
+{
+  routerinfo_t *first = *(routerinfo_t **)a, *second = *(routerinfo_t **)b;
+  tor_addr_t *first_ipv6 = &(first->ipv6_addr);
+  tor_addr_t *second_ipv6 = &(second->ipv6_addr);
+  int comparison = tor_addr_compare(first_ipv6, second_ipv6, CMP_EXACT);
+  // If addresses are equal, use other comparison criterions
+  if (comparison == 0)
+    return compare_routerinfo_usefulness(a, b);
+  else
+    return comparison;
+}
+
+/**
+* Compare routerinfos by descending order of "usefulness" :
+* An authority is more useful than a non-authority; a running router is
+* more useful than a non-running router; and a router with more bandwidth
+* is more useful than one with less.
+**/
+STATIC int
+compare_routerinfo_usefulness(const void **a, const void **b)
```

Comment:
Let's have this function take `const routerinfo_t`, since it is never actually passed to smartlist_sort().